Add option to batch notifications

lilongwe opened the pull request:
When using email for example for notifications it is a bit tedious to receive multiple emails for the same operation. It is less disruptive for messaging like discord or WhatsApp so I have added an option to batch notification in configuration using `BATCH_NOTIFICATIONS` and this will default to false. There is also an option to add a `BATCH_NOTIFICATIONS_TITLE` which has a default of `Batched Notifications` .

I have attached an example for notification to discord and the first one is batched but the second one you can see arrives as multiple messages as it is not batched.

I have added a fix me in `notification.py` as the current implementation creates a new `Apprise` object for every notify and wondering if that can be one object but not familiar with best practises there so left it as it.

![Screenshot 2025-04-02 at 13 17 06](https://github.com/user-attachments/assets/e9349523-da3d-4ee0-a286-3a4d225383b4)

joeShuff commented:
If I understand the code correctly, it only sends the batch when the application quits. 

This works for one-off comparisons but will it work for the scheduled ones where the script continuously runs?

lilongwe commented:
Yeah, good point. Was trying to solve an immediate problem but it breaks
for the the continuous run.

Let me rethink this as there are a few options.

   1. move to running the scheduler with cron and that will work with this
   implementation
   2. add state to the notifications e.g. run_tariff_compare(notifier)
   3. move to using classes which will inherently have state




On Wed, 2 Apr 2025 at 14:19, Joe Shuff ***@***.***> wrote:

> If I understand the code correctly, it only sends the batch when the
> application quits.
>
> This works for one-off comparisons but will it work for the scheduled ones
> where the script continuously runs?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2772543503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIE3JQ4UAQ7XJD7FYRD2XPPVTAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZSGU2DGNJQGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: joeShuff]*joeShuff* left a comment (eelmafia/octopus-minmax#77)
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2772543503>
>
> If I understand the code correctly, it only sends the batch when the
> application quits.
>
> This works for one-off comparisons but will it work for the scheduled ones
> where the script continuously runs?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2772543503>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIE3JQ4UAQ7XJD7FYRD2XPPVTAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZSGU2DGNJQGM>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

joeShuff commented:
Or you could add a method to the `notification.py` to `send_batched_notifications` that gets called at the end of the `run_tariff_compare` function?
<img width="409" alt="image" src="https://github.com/user-attachments/assets/d3ed6dd9-b790-4a9d-9e38-24bcbd6dcd77" />

lilongwe commented:
yes, agreed - i have made the change and it simplifies the code

On Wed, 2 Apr 2025 at 17:45, Joe Shuff ***@***.***> wrote:

> Or you could add a method to the notification.py to
> send_batched_notifications that gets called at the end of the
> run_tariff_compare function?
> image.png (view on web)
> <https://github.com/user-attachments/assets/d3ed6dd9-b790-4a9d-9e38-24bcbd6dcd77>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2773161242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIBMV5YJARDX3IVVXYT2XQH3ZAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZTGE3DCMRUGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
> [image: joeShuff]*joeShuff* left a comment (eelmafia/octopus-minmax#77)
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2773161242>
>
> Or you could add a method to the notification.py to
> send_batched_notifications that gets called at the end of the
> run_tariff_compare function?
> image.png (view on web)
> <https://github.com/user-attachments/assets/d3ed6dd9-b790-4a9d-9e38-24bcbd6dcd77>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2773161242>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIBMV5YJARDX3IVVXYT2XQH3ZAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDONZTGE3DCMRUGI>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>

joeShuff commented on config.py:
```diff
@@ -10,7 +10,11 @@
+BATCH_NOTIFICATIONS = os.getenv("BATCH_NOTIFICATIONS", "false") in ["true", "True", "1"]
+BATCH_NOTIFICATIONS_TITLE = os.getenv("BATCH_NOTIFICATIONS_TITLE", "Batched Notifications")
+
+EXECUTION_TIME = os.getenv("BATCH_NOTIFICATIONS", "23:00")
```
This has the wrong key

lilongwe replied:
that is very strange, seems like a rebase updated that key but fixed it now

lilongwe commented:
@joeShuff bump 😀

joeShuff commented:
I'm happy with this now but I don't have the power to merge PRs so will have to wait for @eelmafia to review

joeShuff commented:
Following the merge of the HA addon, you now need to add support for the new environment variables to the addon config

eelmafia commented:
@lilongwe Good addition, but I want to avoid environment variable bloat. 

Do we really need a `BATCH_NOTIFICATIONS_TITLE`? I think the vast majority won't care enough to customise that. For example, something basic like 'Octopus MinMax Results - <date>' would be fine as a title IMO. Feel free to suggest an alternative title, I just really want to avoid adding extra env variables if we can help it.

lilongwe commented:
Hey

Happy to remove that env variable and just use what you suggested. It keeps
it simple.

Regards

On Mon, 21 Apr 2025 at 21:58, Alex ***@***.***> wrote:

> @lilongwe <https://github.com/lilongwe> Good addition, but I want to
> avoid environment variable bloat.
>
> Do we really need a BATCH_NOTIFICATIONS_TITLE? I think the vast majority
> won't care enough to customise that. For example, something basic like
> 'Octopus MinMax Results - ' would be fine as a title IMO. Feel free to
> suggest an alternative title, I just really want to avoid adding extra env
> variables if we can help it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2819491914>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIEBRVQINUIWSYI7DPL22VLWPAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJZGQ4TCOJRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
> *eelmafia* left a comment (eelmafia/octopus-minmax#77)
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2819491914>
>
> @lilongwe <https://github.com/lilongwe> Good addition, but I want to
> avoid environment variable bloat.
>
> Do we really need a BATCH_NOTIFICATIONS_TITLE? I think the vast majority
> won't care enough to customise that. For example, something basic like
> 'Octopus MinMax Results - ' would be fine as a title IMO. Feel free to
> suggest an alternative title, I just really want to avoid adding extra env
> variables if we can help it.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/eelmafia/octopus-minmax/pull/77#issuecomment-2819491914>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADR7IIEBRVQINUIWSYI7DPL22VLWPAVCNFSM6AAAAAB2JNGC72VHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJZGQ4TCOJRGQ>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

lilongwe commented:
> @lilongwe Good addition, but I want to avoid environment variable bloat.
> 
> Do we really need a `BATCH_NOTIFICATIONS_TITLE`? I think the vast majority won't care enough to customise that. For example, something basic like 'Octopus MinMax Results - <date>' would be fine as a title IMO. Feel free to suggest an alternative title, I just really want to avoid adding extra env variables if we can help it.

Hey @eelmafia I have made the changes to 
- remove the BATCH_NOTIFICATIONS_TITLE env variable
- added BATCH_NOTIFICATIONS to the HA add-on
- added an option to force a notification as some should not be batchable by nature
Please review and let me know if you have any questions.

![Screenshot 2025-04-26 at 12 39 21](https://github.com/user-attachments/assets/0042cbf7-bac8-4f48-8dc9-9dff81591c51)

eelmafia commented:
Looks all good to me. Thanks for the PR ^^
